### PR TITLE
display: Add support for an ST7789V based LCD

### PR DESCRIPTION
**Pull request by marc-cpdesign**
This adds a driver for st7789v lcd controller, and TL019FQV01 lcd.
The bulk of the driver is based on the existing ili9340 driver.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

### CODE REVIEW

**Review by vanwinkeljan — APPROVED** [review]
Looks good to me.

Any chance to also add a sample so that in case of driver changes the sample can be used?

**Comment by marc-cpdesign** [comment]
Thanks for reviewing.!

> Any chance to also add a sample so that in case of driver changes the sample can be used?

I added a small sample, FWIW
I also updated some of the GPIO --> GPIOS in the defines

File: samples/display/st7789v/README.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+| P0.26       | SPI_MISO    |
++-------------+-------------+
+| P0.27       | CS          |
++-------------+-------------+
+| P0.25       | DATA/CMD    |
++-------------+-------------+
+| P0.02       | RESET       |
++-------------+-------------+
+
+You might need to alter these according to your specific board/LCD configuration.
```
Comment (dbkinder) [inline samples/display/st7789v/README.rst]: Can you add the actual example commands to build/flash/run the sample, using the ``.. zephyr-app-commands`` directive. (See other sample docs for an example.

File: samples/display/st7789v/README.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
+.. _st7789v-sample:
+
+ST7789V Display driver
+######################
+
+Overview
+********
+This sample will draw some basic rectangles onto the display.
```
Comment (dbkinder) [inline samples/display/st7789v/README.rst]: Should make a reference to the hardware, assuming this is the right link:
```
This sample draws rectangles onto an `ST7789V display
<https://www.crystalfontz.com/controllers/Sitronix/ST7789V/>.
```
Or is there a link to the specific display module you tested this with?

**Review by dbkinder — CHANGES_REQUESTED** [review]
Looks good, but needs sample build commands.

File: samples/display/st7789v/README.rst (hunk 1 above)
Comment (marc-cpdesign) [reply samples/display/st7789v/README.rst]: Added.

File: samples/display/st7789v/README.rst (hunk 2 above)
Comment (marc-cpdesign) [reply samples/display/st7789v/README.rst]: I added a couple of links, but as I say in the comments, the specific module that I dev/tested this with was a custom board.

File: samples/display/st7789v/sample.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+sample:
+  description: Sample application for ST7789V display
+  name: st7789v_sample
+tests:
+  sample.display.st7789v:
+    build_only: true
```
Comment (vanwinkeljan) [inline samples/display/st7789v/sample.yaml]: Missing `platform_whitelist: nrf52_pca10040`

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]
Thx for the sample on small remark on the sample.yaml (should solve Shippable failures).
Also add a license header to the main.c

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

**Review by vanwinkeljan — APPROVED** [review]

File: drivers/display/Kconfig.st7789v [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+# Copyright (c) 2019 Marc Reilly <marc@cpdesign.com.au>
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig ST7789V
+	bool "ST7789V display driver"
+	depends on SPI
+	help
+	Enable driver for ST7789V display driver.
```
Comment (MaureenHelm) [inline drivers/display/Kconfig.st7789v]: Please indent the help text by two spaces

File: drivers/display/display_st7789v.c [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "display_st7789v.h"
+
+#include <device.h>
+#include <spi.h>
+#include <gpio.h>
+#include <misc/byteorder.h>
+#include <display.h>
```
Comment (MaureenHelm) [inline drivers/display/display_st7789v.c]: Please update the include paths to reflect changes in #16540

File: samples/display/st7789v/boards/nrf52_pca10040.conf [hunk 6]
```diff
@@ -0,0 +1,6 @@
+CONFIG_SPI=y
+CONFIG_GPIO=y
+
+CONFIG_ST7789V=y
+CONFIG_ST7789V_LCD_TL019FQV01=y
+CONFIG_ST7789V_RGB565=y
```
Comment (MaureenHelm) [inline samples/display/st7789v/boards/nrf52_pca10040.conf]: Nothing in this file appears to be specific to the nrf board. Is there a reason you didn't just put it all in the `proj.conf` file?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/display/Kconfig.st7789v (hunk 4 above)
Comment (marc-cpdesign) [reply drivers/display/Kconfig.st7789v]: done

File: samples/display/st7789v/boards/nrf52_pca10040.conf (hunk 6 above)
Comment (marc-cpdesign) [reply samples/display/st7789v/boards/nrf52_pca10040.conf]: move these to proj.conf

File: drivers/display/display_st7789v.c (hunk 5 above)
Comment (marc-cpdesign) [reply drivers/display/display_st7789v.c]: done

**Comment by marc-cpdesign** [comment]
Updated for Maureen's requested changes, rebased to current master

File: drivers/display/display_st7789v.c (hunk 5 above)
Comment (MaureenHelm) [reply drivers/display/display_st7789v.c]: spi, gpio, and display paths also need to be updated

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by zzzh** [comment]
Hello,
I have a ST7789V based LCD module ( a 1.54" one, not TL019FQV01), and nrf52_pca10040 board,  so I did a try on this sample.
Unfortunately,  I got "display_st7789v: Could not get SPI device for LCD", I checked the code and realized that it can't find it from the inited driver.
I don't understand why, any ideas?

**Comment by vanwinkeljan** [comment]
@zzzh could you try to use SPI 1 instead of SPI 0 by changing it in `samples/display/st7789v/nrf52_pca10040.overlay`?

https://github.com/zephyrproject-rtos/zephyr/blob/b631cc09abb4a660b824bffee1e3aba0ab6c7b46/samples/display/st7789v/nrf52_pca10040.overlay#L6-L8

**Comment by zzzh** [comment]
@vanwinkeljan Thanks for replying.
I did not try that, and don't think that's the way to solve it.
Actually, I'd fixed it by adding a line in proj.conf : CONFIG_SPI_x=y, x is the number of SPI channel you chose in the overlay file.

**Pull request merged**